Nifi 1495 - AWS Kinesis Firehose

mans2singh opened the pull request:
This pull request combines Http proxy enhancement for aws processors (nifi-1489) and aws kinesis firehose processor (nifi-1495)

apiri commented:
@mans2singh Would you be able to rebase and squash these commits now that the proxy support has been included?

mans2singh commented:
@apiri I've rebased the forked branch.  Thanks

mans2singh commented:
Just wanted to check if there is any other feedback/recommendation on Kinesis Firehose put processor.

mans2singh commented:
@apiri - Do you have any other comments/suggestions for me ?

apiri commented:
Hi Mans,

Been a bit busy, will certainly get to the full review this week.  Thanks for rebasing.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/AbstractKinesisFirehoseProcessor.java:
```diff
@@ -0,0 +1,89 @@
+            .description("Buffering interval for messages (between 60 and 900 seconds).")
+            .defaultValue("60")
+            .required(false)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
```
I don't see this property being passed through to the associated KinesisFirehoseClient.  Am I overlooking something?

If so, you should prefer the usage of StandardValidators#createLongValidator to create a bounded range.  This would let you then not have to do your customValidate logic.

mans2singh replied:
@apiri - The max buffer size and interval are configuration properties and as you pointed out correctly not used in client.  I've removed them and also have updated validators to use createLongValidator

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/AbstractKinesisFirehoseProcessor.java:
```diff
@@ -0,0 +1,89 @@
+            .sensitive(false)
+            .build();
+
+    public static final PropertyDescriptor MAX_BUFFER_SIZE = new PropertyDescriptor.Builder()
```
Same commentary as above for MAX_BUFFER_INTERVAL.  Not sure how this is being used.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/AbstractKinesisFirehoseProcessor.java:
```diff
@@ -0,0 +1,89 @@
+            .sensitive(false)
+            .build();
+
+    public static final PropertyDescriptor BATCH_SIZE = new PropertyDescriptor.Builder()
```
I see this is used, but again, consider the aforementioned validator to get a range and help remove the unneeded customValidate

mans2singh replied:
@apiri - I've corrected this based on your feedback.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(final ValidationContext validationContext) {
```
With the alternative validators specified above, it should be possible to completely remove the need for this method.

mans2singh replied:
I've removed the custom validator code

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java:
```diff
@@ -0,0 +1,184 @@
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(KINESIS_FIREHOSE_DELIVERY_STREAM_NAME, MAX_BUFFER_INTERVAL,
+                  MAX_BUFFER_SIZE, BATCH_SIZE, REGION, ACCESS_KEY, SECRET_KEY, CREDENTIALS_FILE, AWS_CREDENTIALS_PROVIDER_SERVICE, TIMEOUT,
```
If BUFFER_INTERVAL and BUFFER_SIZE are to go unused, they will also need to be removed from here.

mans2singh replied:
I've removed them (these are configuration params which I mistakenly left in the client code).

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java:
```diff
@@ -0,0 +1,184 @@
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+
+        List<FlowFile> flowFiles = session.get(batchSize);
+        if (flowFiles == null || flowFiles.size() == 0) {
```
Minor:  this will never return null, only an empty collection in the case of there being no input

mans2singh replied:
Corrected per your recommendation.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/test/java/org/apache/nifi/processors/aws/kinesis/firehose/ITPutKinesisFirehose.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Test
+    public void testCustomValidateBatchSize1Valid() {
```
These customValidate tests with the above improvements likely add little and could be removed with adjusted validators.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java:
```diff
@@ -0,0 +1,184 @@
+
+            // Prepare batch of records
+            for (int i = 0; i < flowFiles.size(); i++) {
+                final ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
This section is a bit concerning as we are taking the contents of the FlowFiles and putting these items into memory which is going to impact heap.  According to the Kinesis docs (http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/kinesisfirehose/AmazonKinesisFirehoseClient.html) it appears the maximum value this can support is 1,000 kB raw (before encoding occurs).  Given the default batch size, even if constrained, would be ~250MB/instance of exertion on the heap.  

A couple of points:

- Not overly familiar with the Kinesis API but is there a way to kind of stream these items directly from the session so we don't have to load them into memory.
- It would likely be worthwhile to also inspect each flowfile's size to ensure it does not violate the 1000kB supported.  Doing some quick stepping through, it does not appear this is even detected client side.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/test/java/org/apache/nifi/processors/aws/kinesis/firehose/ITPutKinesisFirehose.java:
```diff
@@ -0,0 +1,149 @@
+/**
+ * This test contains both unit and integration test (integration tests are ignored by default)
+ */
+public class ITPutKinesisFirehose {
```
customValidate tests and similar, if still needed or desired, should go into a standard TestPutKinesisFirehose class so that they are run on each build

mans2singh replied:
Moved unit test to separate class.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/test/java/org/apache/nifi/processors/aws/kinesis/firehose/ITPutKinesisFirehose.java:
```diff
@@ -0,0 +1,149 @@
+     * Comment out ignore for integration tests (requires creds files)
+     */
+    @Test
+    @Ignore
```
The profile to run ITs is not enabled by default and this Ignore is unneeded.  Would prefer to have removed so that no code changes are needed to run them.

mans2singh replied:
Removed ignore annotation for IT tests

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/test/java/org/apache/nifi/processors/aws/kinesis/firehose/ITPutKinesisFirehose.java:
```diff
@@ -0,0 +1,149 @@
+     * Comment out ignore for integration tests (requires creds files)
+     */
+    @Test
+    @Ignore
```
Same as above.

apiri commented:
@mans2singh Did an initial look over the code.  The approach seems good but would like to understand more about the batching process and how we can perform this in a safe manner that does not exhaust heap so readily.  

Also, would you please be able to edit the PR name to something with NIFI-1495 in it so that hopefully the JIRA integration will link and include those items on the issue.  It is unneeded to keep the reference to NIFI-1489 as those changes have been merged and PR closed in #209 

Thanks!

mans2singh commented:
Corrected title of pull request.  Please let me know if there is anything else required.

mans2singh commented:
@apiri Your recommendation for using buffer size is great. So, should we add another property - total batch bytes, and then send them in batches of that size ? If you have any other suggestion/pointers on how this can be implemented, please let me know and I will work on it.  Thanks again for your time.

apiri commented:
@mans2singh Yep.  I think we can probably replace the number of items in a batch size (or make this secondary to the buffer size).  The idea would be to continuously grab FlowFiles until we've reached this threshold and then when the next item arrives that would put us over that limit, we transfer it back to the incoming queue session.transfer(flowFile) **note the lack of relationship**.  We can then create the batch much the same way you have done now.  Does that make sense?

Thanks for your work on this (and all the other great AWS stuff, very popular extensions), and apologies for the lag on following up again on this issue.

mans2singh commented:
@apiri - Just wanted to clarify - We will grab a batch of flow files and check that the flow files combined size threshold is reached.  Then we send the batch to the successful relation, and if not we send the flow files back to the work queue with the transfer method you mentioned.  This might cause additional delays if the threshold set by the user is not reached for some time because the flow files are too small or the rate at which flow files are being input is low.
No worries on the time - I know you guys are busy and I really appreciate your time/advice in helping me work on this project.

apiri commented:
@mans2singh I think your logic for collecting the flowfiles should probably just continuously grab individual files at a time.  This continues until one of a couple scenarios is reached:  max buffer size is met or exceeded (based on the the size of each flowfile that has been collected), there are no more files coming into the processor (one or more flowfiles have been collected but we have not yet eclipsed the max buffer size).  In either case, we would close out that collected batch of files and send them on their way much as you had before.

In terms of the 250MB by default, this is from the default batch size of 250MB and, the worst case scenario, each file is 1MB in size. While each of these files is converted to a byte array for sending, they are continuously sitting on the heap.  In the event multiple instances of this processor are running, we could quickly consume some big chunks of the heap.  What I am proposing is to either get rid of the batch size (that property no longer exists) or make this a secondary consideration where we try to receive a certain batch size but first ensure we do not exceed the configured buffer and second, do not exceed the batch size, with similar semantics to the above scenarios in terms of when those batches are sent on their way plus, when a certain batch size has been reached.

Does that clarify a bit?

mans2singh commented:
@apiri  - To make things simple we get rid of the batch size and we just get one flow file at a time on each invocation of the onTrigger.  We store the flow files in memory until the max size limit or no more are coming in and then send the batch held in memory to kinesis firehose.  Do we have to do anything to "close out" the collected batch ?

apiri commented:
Nope, just send it off to Kinesis and then transfer the files to the appropriate relationship afterward.

apiri commented:
Hey @mans2singh,

Just wanted to see how the progress was going.  We are coming up on release time for 0.6.0 and need to start wrapping things up.  If this is something we can bring to a close in the next day or, great, if not, can we push this off for 0.7.0?

Thanks and let me know!

mans2singh commented:
@aldrin - I've added max buffer size limit check while getting flow files from the session as you had recommended.  I've also kept the batch size limit in place. I've added integration tests to check that the messages are sent once the max buffer size is exceeded. Can you please review the code and let me know if it meets your design ?
Thanks

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java:
```diff
@@ -0,0 +1,180 @@
+            for (int i = 0; i < flowFiles.size(); i++) {
+                FlowFile flowFile = flowFiles.get(i);
+
+                // If flow file too large send it to failure
```
This condition check should still happen before the above buffering considerations.

apiri replied:
On second thought, I am okay with this.  Seeing as how the Firehose also batches/aggregates I do not feel there is any issue with sending batch sizes that may be a bit smaller

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/AbstractKinesisFirehoseProcessor.java:
```diff
@@ -0,0 +1,81 @@
+            .description("Max message buffer size (1-50) MB")
+            .defaultValue("1")
+            .required(false)
+            .addValidator(StandardValidators.createLongValidator(1, 50, true))
```
Likely want to prefer utilization of the DATA_SIZE_VALIDATOR.  Not sure we need to strictly cap this as we are guiding users toward an appropriate path and if they have conditions where they wish to go above the maximum (as well as supporting hardware) that should be fine.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java:
```diff
@@ -0,0 +1,180 @@
+                  PROXY_HOST,PROXY_HOST_PORT));
+
+    /**
+     * Max buffer size 1000kb
```
Now that we have a buffer, should likely call this message size

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java:
```diff
@@ -0,0 +1,180 @@
+        long currentBufferSizeBytes = flowFileCandidate.getSize();
+
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        final int maxBufferSizeBytes = context.getProperty(MAX_MESSAGE_BUFFER_SIZE_MB).asInteger() * 1024 * 1000;
```
In conjunction with the Validator change above, we could also make use of .asDataSize for the property instead of integer.  Will help remove the need for your own conversion.

apiri commented:
@mans2singh Overall, the changes look good and appreciate the inclusion of some additional tests.  Going to do some functional testing but think we can get this merged in.  There are, unfortunately, a lot of conflicts, so for some of the minor points mentioned in the review, would you also be able to please get this rebased so we are good to go for a merge?

Thanks!

apiri commented:
Looking good going to AWS endpoints and services.  Good stuff.

mans2singh commented:
@aldrin - I've moved the check for buffer size and rebased as you had recommended.  I could not find the data size validator that you have mentioned though. Please let me know if you have any other recommendation.  Thanks

mattyb149 commented:
@mans2singh I think Aldrin was referring to DATA_SIZE_VALIDATOR in org.apache.nifi.processor.util.StandardValidators. If you need custom min/max size (in bytes), you can use StandardValidators.createDataSizeBoundsValidator(minBytesInclusive, maxBytesInclusive)

apiri commented:
@mattyb149 has it.

@mans2singh the updates look great.  Concerning the validator, how does 6bed3bcaad61c516eacf268400cdd03c0cd58dae look to you?

Otherwise, I think we can send this one on its way.

mans2singh commented:
@aldrin @mattyb149 Thanks for your pointers and samples for using data validator.  I've made the changes and updated test cases.  Please let me know if there is anything else required.  Thanks again.

apiri commented:
@mans2singh Looks great, thanks for getting those changes together and the contribution.  I'll be merging this into master shortly.
